// @ctrip/cargo-nodejs/lib/cache/cache-config.js
'use strict'

module.exports = {
  /**
   * The max time for cache manager to init
   */
  MAX_INITIAL_TIME: 15000,

  /**
   * The max time for cache manager to init
   */
  MAX_SLAVE_INITIAL_TIME: 20000,

  /**
   * The default delay time when the loader thread loaded nothing.
   */
  DEFAULT_EMPTY_DELAY: 10000,

  /**
   * The default interval when the loader finished.
   */
  DEFAULT_INTERVAL: 5000,

  /**
   * The default delay time when the loader occurred an exception.
   */
  DEFAULT_EXCEPTION_DELAY: 20000,

  /**
   * The default data version.
   */
  DEFAULT_VERSION: '0',

  /**
   * The constant of switch channel
   */
  SWITCH_CHANNEL: 'all',

  /**
   * The name of client's cache.
   */
  ClientCacheName: {
    /**
     * The name of config cache.
     */
    CONFIG: 'client-config',

    /**
     * The name of data cache.
     */
    DATA: 'client-data',

    /**
     * The name of white list data cache.
     */
    DATA_WHITELIST: 'client-data-whitelist',

    /**
     * The name of key cache.
     */
    KEY: 'client-key',

    /**
     * The name of group cache.
     */
    GROUP: 'client-group',

    /**
     * The name of site cache.
     */
    SITE: 'client-site',

    /**
     * The name of source cache.
     */
    SOURCE: 'client-source',

    /**
     * The name of channel cache.
     */
    CHANNEL: 'client-channel',

    /**
     * The name of currency cache.
     */
    CURRENCY: 'client-currency',

    /**
     * The name of locale cache.
     */
    LOCALE: 'client-locale',

    /**
     * The name of switch cache
     */
    SWITCH: 'client-switch',

    /**
     * The name of heartbeat cache.
     */
    HEARTBEAT: 'client-heartbeat'
  },

  /**
   * The type for each cache key.
   */
  KeyType: {
    /**
     * The key type is used for site.
     */
    SITE: 'site',

    /**
     * The key type is used for locale.
     */
    LOCALE: 'locale',

    /**
     * The key type is used for currency.
     */
    CURRENCY: 'currency',

    /**
     * The key type is used for channel.
     */
    CHANNEL: 'channel',

    /**
     * The key type is used for source or device.
     */
    SOURCE: 'source',

    /**
     * The key type is used for group.
     */
    GROUP: 'group',

    /**
     * The key type is used for the data distinguished by appId.
     */
    APPID: 'appId',

    /**
     * The key type is used for key.
     */
    KEY: 'key',

    /**
     * The config type is used for both client and server.
     */
    CONFIG: 'config',

    /**
     * The key type is used for client's data.
     */
    DATA: 'data',

    /**
     * The key type is used for client's white list data.
     */
    DATA_WHITELIST: 'dataWhiteList',

    /**
     * The key type is used for the data that don't need any rules to generate key in the com.ctrip.ibu.cargo.cache.
     */
    ORIGINAL: 'original',

    /**
     * The ip_id is used for the data that can find t_data_publish data in the com.ctrip.ibu.cargo.cache.
     */
    IP_ID: 'ip',

    /**
     * The key type is used for client's switch list data.
     */
    SWITCH: 'switch'
  },

  /**
   * The name of cache events
   */
  CacheEvent: {
    /**
     * The cache of cacheKey has been updated
     */
    UPDATED: 'updated',

    /**
     * The cache of cacheKey has been visited
     */
    VISITED: 'visited',

    /**
     * All caches have been visited
     */
    VISITED_ALL: 'visitedall'
  },

  /**
   * The name of cluster messages
   */
  Message: {
    /**
     *
     */
    INIT: 'inited',
    /**
     *
     */
    UPDATED: 'updated'
  }
}
