// @ctrip/node-vampire-bat/libs/utils/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STAT_EVENT = exports.CONNECT_SET = exports.STATUS = exports.TYPES = exports.UCS = exports.CANARY_REQ = exports.INFOSEC = exports.TRACE_PREFIX = exports.ACT_MINI_VERSION = exports.CAT_TREE_AGGREGATOR = exports.pm2Id = exports.FETCH_SERVER_TIME = exports.SPECIAL_SOURCE_IP_LIST = exports.DEFAULT_PORT = exports.CAT_MERGE_TREE = exports.OVERFLOW_QUEUE_LENGTH = exports.QUEUE_LENGTH = exports.SPAN_TIMEOUT = void 0;
exports.SPAN_TIMEOUT = 30000;
exports.QUEUE_LENGTH = 150; //发送队列的长度
exports.OVERFLOW_QUEUE_LENGTH = 30;
exports.CAT_MERGE_TREE = '_CatMergeTree';
exports.DEFAULT_PORT = 2280;
exports.SPECIAL_SOURCE_IP_LIST = ['10.60.21.4']; //** 信安特殊ip，
exports.FETCH_SERVER_TIME = 120000; //间隔2min获取一次,同java 一致
exports.pm2Id = parseInt(process.env[process.env.instance_var || 'NODE_APP_INSTANCE'], 10) || 0;
exports.CAT_TREE_AGGREGATOR = 'CAT_TREE_AGGREGATOR';
exports.ACT_MINI_VERSION = "018000000"; //判断最小版本号
exports.TRACE_PREFIX = 'x-ctx-';
exports.INFOSEC = 'x-ctx-trace-source';
exports.CANARY_REQ = 'canaryreq';
exports.UCS = {
    'UCS_HEADER_IS_ZONE_REQUEST': 'ucs-is-zone-request',
    'UCS_HEADER_ZONE_CODE': 'ucs-zone-code',
    'UCS_CHECK_FAIL': 'Cat.Ucs.Check.Fail',
    'UCS_ZONE_MISMATCH': 'Zone.Mismatch',
    'UCS_ZONE_NOT_SET': 'Zone.Not.Set'
};
exports.TYPES = {
    SPAN: 'span',
    EVENT: 'event',
    ERROR: 'error',
    TAGS: 'tags',
    HEARTBEAT: 'heartbeat',
    STATMESSAGEEVENT: 'statMessageEvent',
    STATMESSAGE: 'statMessage',
};
exports.STATUS = {
    span: '0',
    event: '0',
    error: 'ERROR',
    tags: '0',
    heartbeat: '0',
    statMessageEvent: '0',
    statMessage: '0'
};
exports.CONNECT_SET = {
    // 是否已经触发了高水位线事件
    highWaterMarkSwitch: false,
    // 设置高水位线为4M，同JAVA一致
    highWaterMark: 4 * 1024 * 1024,
    // 设置低水位线为1M
    lowWaterMark: 1 * 1024 * 1024,
    CONN_TIMEOUT: 5 * 60 * 1000
};
exports.STAT_EVENT = {
    BATCH_FLAG: "*",
    COUNT: "count",
    EQUAL_SIGN: "=",
    BATCH_SPLIT: "&",
    FAIL: "fail",
    MESSAGE_ID: "messageId"
};
