// @ctrip/node-vampire-bat/libs/types/Aggregator/BatStatMsgEvents.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatStatMsgEvents = void 0;
const node_vampire_appconfig_1 = __importDefault(require("@ctrip/node-vampire-appconfig"));
const node_vampire_util_1 = __importDefault(require("@ctrip/node-vampire-util"));
const messageId_1 = require("../../utils/messageId");
const queue_1 = require("../../servlet/queue");
const node_vampire_foundation_framework_1 = __importDefault(require("@ctrip/node-vampire-foundation-framework"));
const const_1 = require("../../utils/const");
const BaseScope_1 = require("../BaseScope");
/**
 * 聚合Tree,无需关联上下文
 */
class BatStatMsgEvents extends BaseScope_1.BaseScope {
    constructor(...args) {
        super(...args);
        this.__properties = {
            idc: node_vampire_foundation_framework_1.default.getServerPropertiesSync("idc"),
            region: node_vampire_foundation_framework_1.default.getServerPropertiesSync("region"),
            az: node_vampire_foundation_framework_1.default.getServerPropertiesSync("az")
        };
        this.__children = [];
        this.__messageId = messageId_1.messageId.getMessageId();
        this.__type = const_1.TYPES.STATMESSAGE;
        this.status = const_1.STATUS[this.__type];
        this.durationInMicro = -1;
        this.__domain = node_vampire_appconfig_1.default['AppID'];
        this.__hostName = process.env.HOSTNAME;
        this.__ipAddress = node_vampire_util_1.default.getHostIp();
        this.__timeStamp = null;
    }
    statMessageTree(type, name, data) {
        return new BatStatMsgEvents(type, name, data);
    }
    end() {
        if (this._itemAggregator) {
            for (const key in this._itemAggregator) {
                if (Object.prototype.hasOwnProperty.call(this._itemAggregator, key)) {
                    const type = key;
                    for (const typeItem in this._itemAggregator[key]) {
                        if (Object.prototype.hasOwnProperty.call(this._itemAggregator[key], typeItem)) {
                            const name = typeItem;
                            const data = this._itemAggregator[key][typeItem];
                            const _statEvent = this.statMessageTree(type, name, null);
                            let str = '' + const_1.STAT_EVENT.BATCH_FLAG + const_1.STAT_EVENT.COUNT + const_1.STAT_EVENT.EQUAL_SIGN + data.count + const_1.STAT_EVENT.BATCH_SPLIT + const_1.STAT_EVENT.BATCH_FLAG + const_1.STAT_EVENT.FAIL + const_1.STAT_EVENT.EQUAL_SIGN + data.error;
                            if (data.messageId) {
                                str += const_1.STAT_EVENT.BATCH_SPLIT + const_1.STAT_EVENT.BATCH_FLAG + const_1.STAT_EVENT.MESSAGE_ID + const_1.STAT_EVENT.EQUAL_SIGN + data.messageId;
                            }
                            _statEvent.data = str;
                            _statEvent.__timeStamp = this.__timeStamp;
                            _statEvent.__type = const_1.TYPES.STATMESSAGEEVENT;
                            this.message && this.message.__children.push(_statEvent);
                        }
                    }
                }
            }
        }
        queue_1.queue.addSendQueue(this);
    }
}
exports.BatStatMsgEvents = BatStatMsgEvents;
