// @ctrip/shark-l10n/es/plural/rules/singular.js
export var DEFAULT = function (integer, fraction, fractionBitsCount) {
    return true;
};
export var EN = function (integer, fraction, fractionBitsCount) {
    return integer === 1 && fractionBitsCount === 0;
};
export var DE = EN;
export var NL = EN;
export var PL = EN;
export var SV = EN;
export var FI = EN;
export var pt_PT = EN;
export var ES = function (integer, fraction, fractionBitsCount) {
    return integer === 1 && fraction === 0;
};
export var EL = ES;
export var NO = ES;
export var KK = ES;
export var IT = ES;
export var FR = function (integer, fraction, fractionBitsCount) {
    return integer === 1 || integer === 0;
};
export var PT = FR;
export var RU = function (integer, fraction, fractionBitsCount) {
    return fractionBitsCount === 0 && integer % 10 === 1 && integer % 100 !== 11;
};
export var TL = function (integer, fraction, fractionBitsCount) {
    var mod10;
    if (fractionBitsCount === 0) {
        mod10 = integer % 10;
    }
    else {
        mod10 = Math.floor(fraction * Math.pow(10, fractionBitsCount)) % 10;
    }
    return mod10 !== 4 && mod10 !== 6 && mod10 !== 9;
};
export var TR = function (integer, fraction, fractionBitsCount) {
    return integer === 1;
};
export var DA = function (integer, fraction, fractionBitsCount) {
    return (integer === 1 && fraction === 0) || (fraction > 0 && (integer === 0 || integer === 1));
};
export var UK = function (integer, fraction, fractionBitsCount) {
    return fractionBitsCount === 0 && integer % 10 === 1 && integer % 100 !== 11;
};
export var AR = function (integer, fraction, fractionBitsCount) {
    return integer === 1 && fraction === 0;
};
