// @ctrip/node-vampire-cache/libs/cache.js
const os = require('os');
const util = require('@ctrip/node-vampire-util');

try {
  let nodeVersion = process.version.split('.')[0].slice(1);
  const prebuildPkg = '@ctrip/node-vampire-cache-prebuild' + nodeVersion + '-' + os.platform() + '-' + os.arch();
  process.dlopen(module, require.resolve(prebuildPkg));
} catch (e) {
  process.dlopen(module, require.resolve('../build/Release/binding.node'));
}

let NativeCache = exports.Cache;

let cacheMap = {};

exports.SIZE_DEFAULT = 6;
exports.SIZE_64 = 6;
exports.SIZE_128 = 7;
exports.SIZE_256 = 8;
exports.SIZE_512 = 9;
exports.SIZE_1K = 10;
exports.SIZE_2K = 11;
exports.SIZE_4K = 12;
exports.SIZE_8K = 13;
exports.SIZE_16K = 14;

function Cache(name, size, blockSize) {
  blockSize = blockSize || exports.SIZE_DEFAULT;
  let md5Name = util.md5(process.version + '_' + os.platform() + '_' + os.arch() + '_' + name + '_' + size + '_' + blockSize).slice(8, 24);
  if (!Object.prototype.hasOwnProperty.call(cacheMap, md5Name)) {
    cacheMap[md5Name] = new NativeCache(md5Name, size, blockSize);
  }
  return cacheMap[md5Name];
}

exports.Cache = Cache;

if (require.main === module && process.argv[2] === 'release') {
  process.argv.slice(3).forEach(exports.release);
}
