// @ctrip/trip-component-utils/es/browser/utils.js
import CargoBrowser from '@ctrip/cargo-nodejs/browser';
import { formatUrlWithTripContext } from '../shared';
export function includes(str, target) {
  return str.indexOf(target) > -1;
}
export var getGroup = function getGroup() {
  if (typeof window === 'undefined') {
    return 'trip';
  }
  var hostname = window.location.hostname || window.location.host;
  return includes(hostname, 'ctrip') ? 'ctrip' : 'trip';
};
export var getTripEnv = function getTripEnv() {
  if (typeof window === 'undefined') {
    return 'prod';
  }
  var hostname = window.location.hostname || window.location.host;
  var find = function find(mayNames, env) {
    for (var index = 0; index < mayNames.length; index++) {
      var name = mayNames[index];
      if (includes(hostname, name)) {
        return env;
      }
    }
  };
  return find(['fws', 'fat', 'dev'], 'fws') || find(['uat'], 'uat') || find(['lpt'], 'lpt') || 'prod';
};
export function getLocaleWithCargoBrowser(opt) {
  var _ref = opt || {},
    isOnline = _ref.isOnline,
    isH5 = _ref.isH5;
  var contextType = isOnline ? 'online' : isH5 ? 'h5' : undefined;
  return CargoBrowser.getLocale(contextType);
}
export function getCurrencyWithCargoBrowser(opt) {
  var _ref2 = opt || {},
    isOnline = _ref2.isOnline,
    isH5 = _ref2.isH5;
  var contextType = isOnline ? 'online' : isH5 ? 'h5' : undefined;
  return CargoBrowser.getCurrency(contextType);
}
export function formatUrlWithCargo(url, opt) {
  if (!url) return '';
  try {
    var _ref3 = opt || {},
      isCover = _ref3.isCover,
      isOnline = _ref3.isOnline,
      isH5 = _ref3.isH5;
    var locale = getLocaleWithCargoBrowser({
      isOnline: isOnline,
      isH5: isH5
    });
    var currency = getCurrencyWithCargoBrowser({
      isOnline: isOnline,
      isH5: isH5
    });
    return formatUrlWithTripContext(url, {
      locale: locale,
      currency: currency
    }, {
      isCover: isCover
    });
  } catch (e) {
    console.error("[TripComponent]: (formatUrlWithCargoLocale) error: ");
    console.error(e);
    return url;
  }
}
