// @ctrip/cargo-nodejs/lib/collector/heart-beat-store.js
'use strict'

class HeartBeatStore {
  constructor (appId, sdkVersion, interval, dataSource, cacheVersion) {
    this.appId = appId
    this.sdkVersion = sdkVersion
    this.interval = interval
    this.dataSource = dataSource
    this.cacheVersion = cacheVersion || ''
    this.data = new Map()
  }

  add (dataId, dataType) {
    const key = dataId + dataType
    if (this.data.has(key)) {
      this.data.get(key).count++
      return
    }
    this.data.set(key, {
      dataId,
      dataType,
      count: 1
    })
  }

  renewInterval (newInterval) {
    this.interval = newInterval
  }

  getDataSize () {
    return this.data.size
  }

  outputStore (latestCacheVersion) {
    if (latestCacheVersion) {
      this.cacheVersion = latestCacheVersion
    }
    const data = Array.from(this.data.values())
    this.data.clear()

    return {
      bean: {
        appId: this.appId,
        dataSource: this.dataSource,
        version: this.sdkVersion,
        duration: this.interval,
        cacheVersion: this.cacheVersion,
        data
      }
    }
  }
}

module.exports = HeartBeatStore
