// @ctrip/trip-component-request/lib/browser/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  request: true,
  createSoaAgent: true,
  jsonp: true,
  trackAjax: true,
  trackAjaxFail: true,
  trackAjaxSuccess: true
};
Object.defineProperty(exports, "createSoaAgent", {
  enumerable: true,
  get: function get() {
    return _soa.createSoaAgent;
  }
});
Object.defineProperty(exports, "jsonp", {
  enumerable: true,
  get: function get() {
    return _jsonp.jsonp;
  }
});
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function get() {
    return _request.request;
  }
});
Object.defineProperty(exports, "trackAjax", {
  enumerable: true,
  get: function get() {
    return _track.trackAjax;
  }
});
Object.defineProperty(exports, "trackAjaxFail", {
  enumerable: true,
  get: function get() {
    return _track.trackAjaxFail;
  }
});
Object.defineProperty(exports, "trackAjaxSuccess", {
  enumerable: true,
  get: function get() {
    return _track.trackAjaxSuccess;
  }
});
var _index = require("../shared/index");
Object.keys(_index).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _index[key];
    }
  });
});
var _request = require("./request");
var _soa = require("./soa");
var _jsonp = require("./jsonp");
var _track = require("./track");
