// @ctrip/trip-component-utils/lib/browser/cookie.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLocaleByCookie = exports.getLocale = exports.getCurrencyByCookie = exports.getCurrency = exports.getCookie = void 0;
exports.getOnlineH5Cookie = getOnlineH5Cookie;
exports.getVid = exports.getUnion = void 0;
var _constant = require("./constant");
var _utils = require("./utils");
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var isH5ByUserAgent = function isH5ByUserAgent() {
  if (typeof navigator === 'undefined') {
    return false;
  }
  var uaLower = navigator.userAgent.toLowerCase();
  var regex = /(ctrip|android|iphone|mobile|webos)/;
  var padRegex = /(ipad|pad)/;
  return regex.test(uaLower) && !padRegex.test(uaLower);
};
var isMobileAgent = function isMobileAgent() {
  var pathname = window.location.pathname;
  var isH5 = /^\/m\//.test(pathname) || isH5ByUserAgent();
  return isH5;
};
var getCookie = function getCookie(cookieName) {
  if (typeof document === 'undefined') {
    return '';
  }
  var cookieStr = document.cookie;
  var reg = new RegExp("".concat(cookieName, "=([^;]+);?"));
  var matched = cookieStr.match(reg);
  return matched ? matched[1] : '';
};
exports.getCookie = getCookie;
function getOnlineH5Cookie(onlineCookie, h5Cookie, def) {
  if (typeof document === 'undefined') {
    return def;
  }
  var isH5 = isMobileAgent();
  var cookieValue = isH5 ? getCookie(h5Cookie) : getCookie(onlineCookie);
  return cookieValue || def;
}
var getLocale = function getLocale() {
  var def = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'en_us';
  var opt = arguments.length > 1 ? arguments[1] : undefined;
  var locale = (0, _utils.getLocaleWithCargoBrowser)(opt);
  if (!locale) {
    return def;
  }
  return locale;
};
exports.getLocale = getLocale;
var getCurrency = function getCurrency() {
  var def = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'USD';
  var opt = arguments.length > 1 ? arguments[1] : undefined;
  var currency = (0, _utils.getCurrencyWithCargoBrowser)(opt);
  if (!currency) {
    return def;
  }
  return currency;
};
exports.getCurrency = getCurrency;
var getLocaleByCookie = function getLocaleByCookie() {
  var def = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'en_us';
  return getOnlineH5Cookie(_constant.COOKIE_ONLINE_LOCALE, _constant.COOKIE_H5_LOCALE, def);
};
exports.getLocaleByCookie = getLocaleByCookie;
var getCurrencyByCookie = function getCurrencyByCookie() {
  var def = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'USD';
  return getOnlineH5Cookie(_constant.COOKIE_ONLINE_CURRENCY, _constant.COOKIE_H5_CURRENCY, def);
};
exports.getCurrencyByCookie = getCurrencyByCookie;
var getVid = function getVid() {
  var bfa = getCookie(_constant.COOKIE_BFA);
  var bfaStrArr = bfa.split('.');
  return bfaStrArr.slice(1, 3).join('.');
};
exports.getVid = getVid;
var getUnion = function getUnion() {
  var unionStr = tryDecodeURIComponent(getCookie(_constant.COOKIE_UNION));
  if (unionStr) {
    var unionData = parseQuery(unionStr);
    return {
      aid: unionData.AllianceID || '',
      sid: unionData.SID || ''
    };
  }
  return {
    aid: '',
    sid: ''
  };
};
exports.getUnion = getUnion;
function parseQuery(query) {
  var out = {};
  query.split('&').forEach(function (item) {
    var _item$split = item.split('='),
      _item$split2 = _slicedToArray(_item$split, 2),
      key = _item$split2[0],
      _item$split2$ = _item$split2[1],
      val = _item$split2$ === void 0 ? '' : _item$split2$;
    var decodeVal = tryDecodeURIComponent(val);
    out[key] = decodeVal;
  });
  return out;
}
function tryDecodeURIComponent(str) {
  if (!str) {
    return '';
  }
  try {
    return decodeURIComponent(str);
  } catch (error) {
    return str;
  }
}
