// @ctrip/trip-component-utils/lib/browser/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  getTripEnv: true,
  getGroup: true,
  formatUrlWithCargo: true,
  getCookie: true,
  getVid: true,
  getUnion: true,
  getLocale: true,
  getCurrency: true
};
Object.defineProperty(exports, "formatUrlWithCargo", {
  enumerable: true,
  get: function get() {
    return _utils.formatUrlWithCargo;
  }
});
Object.defineProperty(exports, "getCookie", {
  enumerable: true,
  get: function get() {
    return _cookie.getCookie;
  }
});
Object.defineProperty(exports, "getCurrency", {
  enumerable: true,
  get: function get() {
    return _cookie.getCurrency;
  }
});
Object.defineProperty(exports, "getGroup", {
  enumerable: true,
  get: function get() {
    return _utils.getGroup;
  }
});
Object.defineProperty(exports, "getLocale", {
  enumerable: true,
  get: function get() {
    return _cookie.getLocale;
  }
});
Object.defineProperty(exports, "getTripEnv", {
  enumerable: true,
  get: function get() {
    return _utils.getTripEnv;
  }
});
Object.defineProperty(exports, "getUnion", {
  enumerable: true,
  get: function get() {
    return _cookie.getUnion;
  }
});
Object.defineProperty(exports, "getVid", {
  enumerable: true,
  get: function get() {
    return _cookie.getVid;
  }
});
var _index = require("../shared/index");
Object.keys(_index).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _index[key];
    }
  });
});
var _utils = require("./utils");
var _cookie = require("./cookie");
