// @ctrip/trip-component-utils/lib/browser/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatUrlWithCargo = formatUrlWithCargo;
exports.getCurrencyWithCargoBrowser = getCurrencyWithCargoBrowser;
exports.getGroup = void 0;
exports.getLocaleWithCargoBrowser = getLocaleWithCargoBrowser;
exports.getTripEnv = void 0;
exports.includes = includes;
var _browser = _interopRequireDefault(require("@ctrip/cargo-nodejs/browser"));
var _shared = require("../shared");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function includes(str, target) {
  return str.indexOf(target) > -1;
}
var getGroup = function getGroup() {
  if (typeof window === 'undefined') {
    return 'trip';
  }
  var hostname = window.location.hostname || window.location.host;
  return includes(hostname, 'ctrip') ? 'ctrip' : 'trip';
};
exports.getGroup = getGroup;
var getTripEnv = function getTripEnv() {
  if (typeof window === 'undefined') {
    return 'prod';
  }
  var hostname = window.location.hostname || window.location.host;
  var find = function find(mayNames, env) {
    for (var index = 0; index < mayNames.length; index++) {
      var name = mayNames[index];
      if (includes(hostname, name)) {
        return env;
      }
    }
  };
  return find(['fws', 'fat', 'dev'], 'fws') || find(['uat'], 'uat') || find(['lpt'], 'lpt') || 'prod';
};
exports.getTripEnv = getTripEnv;
function getLocaleWithCargoBrowser(opt) {
  var _ref = opt || {},
    isOnline = _ref.isOnline,
    isH5 = _ref.isH5;
  var contextType = isOnline ? 'online' : isH5 ? 'h5' : undefined;
  return _browser.default.getLocale(contextType);
}
function getCurrencyWithCargoBrowser(opt) {
  var _ref2 = opt || {},
    isOnline = _ref2.isOnline,
    isH5 = _ref2.isH5;
  var contextType = isOnline ? 'online' : isH5 ? 'h5' : undefined;
  return _browser.default.getCurrency(contextType);
}
function formatUrlWithCargo(url, opt) {
  if (!url) return '';
  try {
    var _ref3 = opt || {},
      isCover = _ref3.isCover,
      isOnline = _ref3.isOnline,
      isH5 = _ref3.isH5;
    var locale = getLocaleWithCargoBrowser({
      isOnline: isOnline,
      isH5: isH5
    });
    var currency = getCurrencyWithCargoBrowser({
      isOnline: isOnline,
      isH5: isH5
    });
    return (0, _shared.formatUrlWithTripContext)(url, {
      locale: locale,
      currency: currency
    }, {
      isCover: isCover
    });
  } catch (e) {
    console.error("[TripComponent]: (formatUrlWithCargoLocale) error: ");
    console.error(e);
    return url;
  }
}
