// @ctrip/cargo-nodejs/lib/service/abstract-client.js
'use strict'

class AbstractClient {
  getGroupData (version, force = true) {
    throw new Error('groupData')
  }

  getSiteData (version, force = true) {
    throw new Error('siteData')
  }

  getLocaleData (version, force = true) {
    throw new Error('localeData')
  }

  getSourceData (version, force = true) {
    throw new Error('sourceData')
  }

  getCurrencyData (version, force = true) {
    throw new Error('currencyData')
  }

  getChannelData (version, force = true) {
    throw new Error('channelData')
  }

  getKeyData (version, force = true) {
    throw new Error('keyData')
  }

  getAppKeyData (appId, version, force = true) {
    throw new Error('appKeyData')
  }

  getConfigData (appId) {
    throw new Error('configData')
  }

  getData (appId, version, force = true) {
    throw new Error('data')
  }

  getAllData (appId, version, force = true) {
    throw new Error('allData')
  }

  getDataWhiteList (appId, version, force = true) {
    throw new Error('dataWhiteList')
  }

  getCurrencySwitch (force = true) {
    throw new Error('currencySwitch')
  }

  sendHeartBeat (heartBeatData) {
    return new Error('sendHeart')
  }
}

module.exports = AbstractClient
