// @ctrip/shark-l10n/es/number/template.js
export var DEFAULT_NUMBER_TEMPLATE = '${value}';
export var MAX_SHORT_NUMBER_POWER = 12;
export var DEFAULT_INVALID_FACTOR = '-1';
export var UNICODE_SYMBOL = {
    APOSTROPHE: unescape('%27'),
    COMMA: unescape('%2C'),
    DOT_ABOVE: unescape('%u02D9'),
    FULL_STOP: '.',
    MIDDLE_DOT: unescape('%B7'),
    NARROW_NO_BREAK_SPACE: unescape('%u202F'),
    SPACE: unescape('%20'),
    THIN_SPACE: unescape('%u2009'),
};
export var TEMPLATE_MAP = {
    CURRENCY_NUMBER: function (sym, value) { return "".concat(sym).concat(value); },
    CURRENCY_SPACE_NUMBER: function (sym, value) { return "".concat(sym, " ").concat(value); },
    NUMBER_CURRENCY: function (sym, value) { return "".concat(value).concat(sym); },
    NUMBER_SPACE_CURRENCY: function (sym, value) { return "".concat(value, " ").concat(sym); },
};
export var SHORT_NUMBER_TEMPLATE_MAP = {
    NUMBER_SPACE_SYMBOL: function (sym, value) { return "".concat(value, " ").concat(sym); },
    NUMBER_SYMBOL: function (sym, value) { return "".concat(value).concat(sym); },
    SYMBOL_NUMBER: function (sym, value) { return "".concat(sym).concat(value); },
    SYMBOL_SPACE_NUMBER: function (sym, value) { return "".concat(sym, " ").concat(value); },
};
