// @ctrip/trip-component-request/lib/shared/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  RequestError: true,
  isRequestError: true,
  TimeoutError: true,
  isTimeoutError: true,
  AbortError: true,
  isAbortError: true,
  setDefaultErrorHandler: true
};
Object.defineProperty(exports, "AbortError", {
  enumerable: true,
  get: function get() {
    return _errors.AbortError;
  }
});
Object.defineProperty(exports, "RequestError", {
  enumerable: true,
  get: function get() {
    return _errors.RequestError;
  }
});
Object.defineProperty(exports, "TimeoutError", {
  enumerable: true,
  get: function get() {
    return _errors.TimeoutError;
  }
});
Object.defineProperty(exports, "isAbortError", {
  enumerable: true,
  get: function get() {
    return _errors.isAbortError;
  }
});
Object.defineProperty(exports, "isRequestError", {
  enumerable: true,
  get: function get() {
    return _errors.isRequestError;
  }
});
Object.defineProperty(exports, "isTimeoutError", {
  enumerable: true,
  get: function get() {
    return _errors.isTimeoutError;
  }
});
Object.defineProperty(exports, "setDefaultErrorHandler", {
  enumerable: true,
  get: function get() {
    return _defaultErrorHandler.setDefaultErrorHandler;
  }
});
var _utils = require("./utils");
Object.keys(_utils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _utils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _utils[key];
    }
  });
});
var _errors = require("./errors");
var _defaultErrorHandler = require("./default-error-handler");
