// @ctrip/node-vampire-triplog/libs/utils/commonConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LIMIT_RANGE = exports.TRIPLOGGING_MESSAGE_TYPES = exports.constants = void 0;
exports.constants = {
    KEY_HOST_ADDRESS: '__hostAddress',
    KEY_HOSTNAME: '__hostName',
    KEY_ZONE: '__zone',
    KEY_GROUP: '__group',
    KEY_SUBENV: '__subEnv',
    KEY_LOG_SCENARIO: '__scenario',
    KEY_EXCEPTION: '__exception',
    KEY_CAT_MARKER_ID: 'cat-marker-id',
    KEY_CAT_MESSAGE_ID: 'cat-msg-id',
    LABEL_DEFAULT: 'default',
    SCENARIOS_KEYS: 'scenario.keys',
    KEY_CLOG_AZ: '__az',
    KEY_CLOG_REGION: '__region',
    KEY_CLOG_IDC: '__idc'
};
exports.TRIPLOGGING_MESSAGE_TYPES = ['debug', 'info', 'warn', 'error', 'trace'];
/**限制性参数 */
exports.LIMIT_RANGE = {
    DEFAULT_MAX_TAG_TITLE_LENGTH: 200,
    DEFAULT_MAX_TAG_KEY_LENGTH: 512 * 1024,
    DEFAULT_MAX_TAG_VALUE_LENGTH: 16 * 1024 * 1024,
    DEFAULT_MAX_MESSAGE_LENGTH: 16 * 1024 * 1024,
    MAX_RETRY_ATTEMPTS: 2,
    maxBatchSize: 16777216,
    maxBatchCount: 100
};
