// @ctrip/node-vampire-metric/appConfig.js
const appConfigModule = require("@ctrip/node-vampire-appconfig");
const frameworkFoundation = require("@ctrip/node-vampire-foundation-framework");
const request = require("@ctrip/node-vampire-http");
const log=require('./console');

class MetricConfig {
  constructor() {
    this.loggingServer_Bootstrap = {
      fws: "http://clog.bootstrap.fx.fws.qa.nt.ctripcorp.com/api/bootstrap",
      lpt: "http://clog.bootstrap.fx.fws.qa.nt.ctripcorp.com/api/bootstrap",
      uat: "http://clog.bootstrap.fx.uat.qa.nt.ctripcorp.com/api/bootstrap",
      pro: "http://clog.bootstrap.fx.ctripcorp.com/api/bootstrap"
    }
    this.loggingServer_V2_IP = {
      fws: "collector.logging.fws.qa.nt.ctripcorp.com",
      lpt: "collector.logging.fws.qa.nt.ctripcorp.com",
      uat: "collector.logging.uat.qa.nt.ctripcorp.com",
      pro: "collector.logging.ctripcorp.com",
    };
    this.debug = appConfigModule["debug"];
    this.hookNoStack = appConfigModule["hookNoStack"];
    this.setDefaultConfig("LoggingServer.V2.Port", 63100);
    this.interval = null;
    this.init();
  }

  setDefaultConfig(key, value) {
    this[key] = value;
    appConfigModule[key] = value;
  }

  updateMetricServerConfig() {
    this.setDefaultConfig(
      "LoggingServer.V2.IP",
      this.loggingServer_V2_IP[this.env]
    );
    this.setDefaultConfig(
      "LoggingServer.Bootstrap",
      this.loggingServer_Bootstrap[this.env]
    );
  }

  init() {
    if (!this.initPromise) {
      this.initPromise = new Promise((resolve, reject) => {
        Promise.all(
          [frameworkFoundation.getEnv(), frameworkFoundation.getIDC(), frameworkFoundation.getZone()]
        ).then((args) => {
          let env = args[0];
          let idc = args[1];
          let zone = args[2];

          this.AppID = frameworkFoundation.getAppID();
          this.env = env;
          this.IDC = idc;
          this.zone = zone;
          this.setDefaultConfig("LogLevel", 1);

          this.updateMetricServerConfig();
          const loadServerList = () => {
            request(appConfigModule["LoggingServer.Bootstrap"], {
              "contentType": "json",
              "unref":true,
              "responseTimeout":60000
            }, JSON.stringify({
              idc: this.IDC
            })).then((result) => {
              if (result.bootstrapServers) {
                this.setDefaultConfig(
                  "LoggingServer.Bootstrap.IPList",
                  result.bootstrapServers
                );
                this.setDefaultConfig("LoggingServer.Bootstrap.Port", result.port);
              }
            }).catch(() => {
              // empty
            });
          };
          /**2019/6/18修改 判断是否读取metadata成功，如果成功的话，则直接覆盖
           * 不再hard-code env和idc
           */
          frameworkFoundation
            .getConfig("*")
            .then((data) => {
              if (frameworkFoundation.configNeedToBeUpdated) {
                if (data["clog.bootstrap"]) {
                  appConfigModule["LoggingServer.Bootstrap"] = data["clog.bootstrap"];
                }
                if (data["clog.servers"]) {
                  const logServers = data["clog.servers"].split(',');
                  appConfigModule["LoggingServer.V2.IP"] = logServers && logServers[Math.floor(Math.random() * logServers.length)];
                }
                if (data["clog.log.transfertocat.enable"]) {
                  appConfigModule["CLOG_LOG_TRANSFERTOCAT_ENABLE"] =
                    data["clog.log.transfertocat.enable"];
                }
              }
              loadServerList();
              return resolve();
            }).catch((e) => {
              //如果没有metadata_domain没有返回，则仍然返回默认设置的（数据无法回发，但是不会影响应用运行）
              log(
                "Get metric.server.host error, reuse the default server value",
                e
              );
              return resolve();
            });
          this.interval = setInterval(() => {
            loadServerList();
          }, 60000).unref();
        }).catch((e) => {
          log('Init config error, ', e);
          delete this.initPromise;
          clearInterval(this.interval);
          return reject(e);
        });
      });
    }
    return this.initPromise
  }
}

module.exports = new MetricConfig();
