// @ctrip/cargo-nodejs/lib/context/resolver/resolver.js
'use strict'

class Resolver {
  /**
   * @param {string} groupName
   * @param {string} siteName
   * @param {string} contextType
   * @param {boolean} h5UseSiteForce
   */
  constructor (groupName, siteName, contextType, h5UseSiteForce) {
    this.groupName = groupName
    this.siteName = siteName
    this.contextType = contextType
    this.supportedValues = this.getSupportedValues(h5UseSiteForce)
  }

  /**
   * get default value
   *
   * @return {string} default value
   */
  getDefault () {
    const recommendValues = this.getRecommendValues()

    for (const value of recommendValues) {
      if (this.getNormalizedValue(value) !== false) {
        return value
      }
    }

    return this.getGlobalDefault()
  }

  /**
   * if the target is supported, return normalized value, else return false
   *
   * @param {string} target
   * @return {boolean|string}
   */
  getNormalizedValue (target) {
    if (!target) {
      return false
    }

    for (const value of this.supportedValues) {
      if (value.toLowerCase() === target.toLowerCase()) {
        return value
      }
    }

    return false
  }

  /**
   * get global default value
   * @return {string}
   */
  getGlobalDefault () {
    throw new Error('get global default')
  }

  /**
   * @param {boolean} h5UseSiteForce
   * get supported locales / currencies
   * @return {string []}
   */
  getSupportedValues (h5UseSiteForce) {
    throw new Error('get supported locales / currencies')
  }

  /**
   * get recommend values for default
   * @return {string}
   */
  getRecommendValues () {
    throw new Error('get recommend values for default')
  }
}

module.exports = Resolver
