// @ctrip/node-vampire-bat/libs/types/BaseScope.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseScope = void 0;
const node_vampire_util_1 = __importDefault(require("@ctrip/node-vampire-util"));
class BaseScope {
    constructor(type, name, data) {
        this.time = new Date().getTime();
        this.__children = null;
        this.__startMicroSeconds = node_vampire_util_1.default.getMicroSeconds();
        this.__close = false;
        this.__properties = {};
        this._uid = node_vampire_util_1.default.uid();
        this.name = name || 'Unknow Name';
        this.type = type || 'Unknow Type';
        this.data = data || '';
        this.__startTime = Date.now();
    }
    baseBeforeSend() {
        if (this.__properties["Tree.Begin"]) {
            this.__properties["Tree.End"] = new Date().getTime();
        }
        this.__close = true;
    }
    debug() {
        return `${this.__type} ${node_vampire_util_1.default.formatDateTime(this.__startTime, "hh:mm:ss:SSS")} ${this.__messageId} ${this.type} ${this.name} ${this.truncateString(this.data, 50)}`;
    }
    truncateString(str, len) {
        if (/\r?\n/.test(str)) {
            try {
                str = JSON.stringify(JSON.parse(str));
            }
            catch (e) {
                str = str.replace(/(\r?\n)+/g, ' ');
            }
        }
        if (str.length > len) {
            str = str.slice(0, len) + '...[' + str.length + ']';
        }
        return str;
    }
}
exports.BaseScope = BaseScope;
