// @ctrip/shark-l10n/lib/plural/number.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluralNum = void 0;
var PluralNum = (function () {
    function PluralNum(num) {
        var arr = this.toString(num).split('.');
        this.integer = parseInt(arr[0], 10);
        this.fraction = 0;
        this.fractionBitsCount = 0;
        if (arr.length > 1) {
            this.fraction = Number('.' + arr[1]);
            this.fractionBitsCount = arr[1].length;
        }
        if (this.fractionBitsCount > 7) {
            this.fraction = Number(this.fraction.toFixed(5));
            this.fractionBitsCount = 5;
        }
    }
    PluralNum.prototype.toString = function (num) {
        if (typeof num === 'string') {
            return num;
        }
        var numNativeString = num.toString();
        if (Math.abs(num) < 1.0) {
            var e = parseInt(numNativeString.split('e-')[1], 10);
            if (e) {
                num *= Math.pow(10, e - 1);
                num = '0.' + (new Array(e)).join('0') + numNativeString.substring(3);
                return num;
            }
        }
        else {
            var e = parseInt(numNativeString.split('+')[1], 10);
            if (e > 20) {
                e -= 20;
                num /= Math.pow(10, e);
                num += (new Array(e + 1)).join('0');
                return num;
            }
        }
        return numNativeString;
    };
    PluralNum.prototype.getInteger = function () {
        return this.integer;
    };
    PluralNum.prototype.getFration = function () {
        return this.fraction;
    };
    PluralNum.prototype.getFractionBitsCount = function () {
        return this.fractionBitsCount;
    };
    return PluralNum;
}());
exports.PluralNum = PluralNum;
