// @ctrip/shark-l10n/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPlural = exports.PhoneFormatter = exports.NumberFormatter = exports.NameFormatter = exports.MeasurementFormatter = exports.DurationFormatter = exports.DateTimeFormatter = exports.L10n = void 0;
var datetime_1 = require("./datetime");
Object.defineProperty(exports, "DateTimeFormatter", { enumerable: true, get: function () { return datetime_1.DateTimeFormatter; } });
var duration_1 = require("./duration");
Object.defineProperty(exports, "DurationFormatter", { enumerable: true, get: function () { return duration_1.DurationFormatter; } });
var measurement_1 = require("./measurement");
Object.defineProperty(exports, "MeasurementFormatter", { enumerable: true, get: function () { return measurement_1.MeasurementFormatter; } });
var name_1 = require("./name");
Object.defineProperty(exports, "NameFormatter", { enumerable: true, get: function () { return name_1.NameFormatter; } });
var number_1 = require("./number");
Object.defineProperty(exports, "NumberFormatter", { enumerable: true, get: function () { return number_1.NumberFormatter; } });
var phone_1 = require("./phone");
Object.defineProperty(exports, "PhoneFormatter", { enumerable: true, get: function () { return phone_1.PhoneFormatter; } });
var plural_1 = require("./plural");
Object.defineProperty(exports, "getPlural", { enumerable: true, get: function () { return plural_1.getPlural; } });
__exportStar(require("./types"), exports);
var L10n = (function () {
    function L10n(i18n) {
        this.setI18n(i18n);
    }
    L10n.prototype.setI18n = function (i18n) {
        this.i18n = i18n;
        return this;
    };
    L10n.prototype.DateTimeFormat = function (timezone) {
        return new datetime_1.DateTimeFormatter(this.i18n, timezone);
    };
    L10n.prototype.DurationFormat = function (locale) {
        return new duration_1.DurationFormatter(this.i18n, locale);
    };
    L10n.prototype.MeasurementFormat = function () {
        return new measurement_1.MeasurementFormatter(this.i18n);
    };
    L10n.prototype.NameFormat = function () {
        return new name_1.NameFormatter(this.i18n);
    };
    L10n.prototype.NumberFormat = function () {
        return new number_1.NumberFormatter(this.i18n);
    };
    L10n.prototype.PhoneFormat = function () {
        return new phone_1.PhoneFormatter(this.i18n);
    };
    L10n.prototype.getPlural = function (key, locale, num) {
        return (0, plural_1.getPlural)(this.i18n, key, locale, num);
    };
    return L10n;
}());
exports.L10n = L10n;
