// @ctrip/shark-sdk/lib/log/impl/dev-logger.js
'use strict'
const pm2 = require('@ctrip/node-vampire-pm2')
const is = require('is-type-of')
const colors = require('colors/safe')
const Logger = require('../logger')

class DevLogger extends Logger {
  get enabledEnv () {
    return ['', 'dev', 'fat', 'fws', 'uat', 'lpt']
  }

  common () {
    const common = [colors.green('[shark]')]
    common.push(colors.green(pm2.isMaster ? '<master>' : '<slave>'))
    return common.join(' ')
  }

  parseArgs (args) {
    if (is.isString(args)) {
      return args
    } else if (is.isArray(args)) {
      return args.map(arg => this.parseArgs(arg)).join(' | ')
    } else if (is.isError(args)) {
      return `${args.toString()}\n${args.stack}`
    } else if (is.isObject(args)) {
      const values = []
      for (const k in args) {
        let value = args[k]
        if (is.isObject(value)) {
          value = JSON.stringify(args[k])
        }
        values.push(`${k} => ${value}`)
      }
      return values.join(', ')
    } else {
      return String(args)
    }
  }

  debug (args) {
    const logs = this.parseArgs(args)
    if (!logs) {
      return
    }
    const output = this.common() + colors.cyan(' DEBUG ') + logs
    console.log(output)
  }

  info (args) {
    const logs = this.parseArgs(args)
    if (!logs) {
      return
    }
    const output = this.common() + colors.green(' INFO ') + logs
    console.log(output)
  }

  warning (args) {
    const logs = this.parseArgs(args)
    if (!logs) {
      return
    }
    const output = this.common() + colors.yellow(' WARNING ') + logs
    console.log(output)
  }

  error (args) {
    const logs = this.parseArgs(args)
    if (!logs) {
      return
    }
    const output = this.common() + colors.red(' ERROR ') + logs
    console.log(output)
  }
}

module.exports = new DevLogger()
