// @ctrip/cargo-nodejs/lib/context/context-config.js
module.exports = {
  /**
   * The key of locale in query string.
   */
  QUERY_STRING_LOCALE_KEY: 'locale',

  /**
   * The key of locale in http header.
   */
  HEADER_LOCALE_KEY: 'locale',

  /**
   * The key of locale in cookie.
   */
  COOKIE_LOCALE_KEY: 'ibulocale',

  /**
   * The key of currency in query string.
   */
  QUERY_STRING_CURRENCY_KEY: 'curr',

  /**
   * The key of currency alias in query string.
   */
  QUERY_STRING_CURRENCY_KEY_ALIAS: 'currency',

  /**
   * The key of currency in http header.
   */
  HEADER_CURRENCY_KEY: 'currency',

  /**
   * The key of currency in cookie.
   */
  COOKIE_CURRENCY_KEY: 'cookiePricesDisplayed',

  /**
   * The key of language in query string.
   */
  QUERY_STRING_LANGUAGE_KEY: 'language',

  /**
   * The key of language alias in query string.
   */
  QUERY_STRING_LANGUAGE_KEY_ALIAS: 'lang',

  /**
   * The key of language in http header.
   */
  HEADER_LANGUAGE_KEY: 'language',

  /**
   * The key of language in cookie.
   */
  COOKIE_LANGUAGE_KEY: 'ibulanguage',

  /**
   * The names of universal(online&h5) cookies
   */
  UNIVERSAL_COOKIES: {
    CURRENCY: 'ibu_currency',
    LANGUAGE: 'ibu_language',
    LOCALE: 'ibu_locale',
    GROUP: 'ibu_group',
    SITE: 'ibu_site'
  },

  /**
   * The names of H5 Cookies
   */
  H5_COOKIES: {
    CURRENCY: 'ibu_h5_curr',
    LANGUAGE: 'ibu_h5_lang',
    LOCALE: 'ibu_h5_local',
    GROUP: 'ibu_h5_group',
    SITE: 'ibu_h5_site'
  },

  /**
   * The key of host in http header.
   */
  HOST_KEY: 'host',

  /**
   * The separator for getting host in http header.
   */
  SEPARATOR_HOST: ':',

  /**
   * The default domain.
   */
  DEFAULT_DOMAIN: 'trip.com',

  /**
   * The default currency
   */
  DEFAULT_CURRENCY: 'USD',

  /**
   * The default locale.
   */
  DEFAULT_LOCALE: 'en-US',

  /**
   * The default language
   */
  DEFAULT_LANGUAGE: 'en',

  /**
   * The name in table key labels frequently used currencies.
   */
  KEY_NAME_TOP_CURRENCIES: 'TopCurrencies',

  /**
   * The name in table key labels frequently used locales.
   */
  KEY_NAME_SUPPORTED_LOCALES: 'SupportedLocales',

  /**
   * The name in table key labels cookie domain.
   */
  KEY_NAME_COOKIE_DOMAIN: 'CookieDomains',

  /**
   * The name in table key labels cookie domain.
   */
  KEY_NAME_COOKIE_POLICY: 'CookiePolicy',

  /**
   * The name in table key labes locale alias
   */
  KEY_NAME_LOCALE_ALIAS: 'LocaleAlias',

  /**
   * The name in table key labels locale language
   */
  KEY_NAME_LOCALE_LANGUAGE: 'LocaleLanguage',

  /**
   * The name in table key labels locale csharp
   */
  KEY_NAME_LOCALE_CSHARP: 'LocaleCSharp',

  /**
   * The name in table key labels h5 site name
   */
  KEY_NAME_H5_SITENAME: 'H5SiteName',

  /**
   * The enum for ContextType
   */
  CONTEXT_TYPE: {

    H5: 'h5',

    ONLINE: 'online'

  }
}
