// @ctrip/trip-component-request/lib/node/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("../shared/index"), exports);
var request_1 = require("./request");
Object.defineProperty(exports, "request", { enumerable: true, get: function () { return request_1.request; } });
var soa_1 = require("./soa");
Object.defineProperty(exports, "createSoaAgent", { enumerable: true, get: function () { return soa_1.createSoaAgent; } });
var jsonp_1 = require("./jsonp");
Object.defineProperty(exports, "jsonp", { enumerable: true, get: function () { return jsonp_1.jsonp; } });
var track_1 = require("./track");
Object.defineProperty(exports, "trackAjax", { enumerable: true, get: function () { return track_1.trackAjax; } });
Object.defineProperty(exports, "trackAjaxFail", { enumerable: true, get: function () { return track_1.trackAjaxFail; } });
Object.defineProperty(exports, "trackAjaxSuccess", { enumerable: true, get: function () { return track_1.trackAjaxSuccess; } });
