// is-type-of/dist/commonjs/types/primitive.js
"use strict";
// Javascript Primitive
// https://developer.mozilla.org/en-US/docs/Glossary/Primitive
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFinite = exports.isNaN = exports.isDouble = exports.isSafeInteger = exports.isLong = exports.isInteger32 = exports.isInteger = exports.isPrimitive = exports.isNullable = exports.isNull = exports.isUndefined = exports.isSymbol = exports.isBigInt = exports.isBoolean = exports.isNumber = exports.isString = void 0;
const MAX_INT_31 = Math.pow(2, 31);
/**
 * returns true if val is string, not `String` object
 */
function isString(val) {
    return typeof val === 'string';
}
exports.isString = isString;
/**
 * returns true if val is number, not `Number` object
 */
function isNumber(val) {
    return typeof val === 'number';
}
exports.isNumber = isNumber;
/**
 * returns true if val is boolean, not `Boolean` object
 */
function isBoolean(val) {
    return typeof val === 'boolean';
}
exports.isBoolean = isBoolean;
/**
 * returns true if val is bigint
 */
function isBigInt(val) {
    return typeof val === 'bigint';
}
exports.isBigInt = isBigInt;
/**
 * returns true if val is symbol
 */
function isSymbol(val) {
    return typeof val === 'symbol';
}
exports.isSymbol = isSymbol;
/**
 * returns true if val is undefined
 */
function isUndefined(val) {
    return typeof val === 'undefined';
}
exports.isUndefined = isUndefined;
/**
 * returns true if val is null
 */
function isNull(val) {
    return val === null;
}
exports.isNull = isNull;
/**
 * returns true if val is null or undefined
 */
function isNullable(val) {
    return val == null;
}
exports.isNullable = isNullable;
/**
 * returns true if val is primitive
 */
function isPrimitive(val) {
    return val === null ||
        typeof val === 'boolean' ||
        typeof val === 'number' ||
        typeof val === 'string' ||
        typeof val === 'symbol' ||
        typeof val === 'bigint' ||
        typeof val === 'undefined';
}
exports.isPrimitive = isPrimitive;
// Number Extension
/**
 * returns true if val is integer
 */
function isInteger(val) {
    if (isNullable(val))
        return false;
    return Number.isInteger(val);
}
exports.isInteger = isInteger;
/**
 * returns true if val is integer, and between -2 ** 31 and 2 ** 31 - 1
 */
function isInteger32(val) {
    if (!isInteger(val))
        return false;
    return val >= -MAX_INT_31 && val <= MAX_INT_31 - 1;
}
exports.isInteger32 = isInteger32;
/**
 * returns true if val is integer, and < -2 ** 31, and > 2 ** 31 - 1
 */
function isLong(val) {
    if (!isInteger(val))
        return false;
    return val < -MAX_INT_31 || val > MAX_INT_31 - 1;
}
exports.isLong = isLong;
/**
 * returns true if val is integer, and between -(2 ** 53 - 1) and 2 ** 53 - 1
 * @see {Number.isSafeInteger}
 */
function isSafeInteger(val) {
    return Number.isSafeInteger(val);
}
exports.isSafeInteger = isSafeInteger;
/**
 * returns true if val is Double
 */
function isDouble(val) {
    if (isNullable(val))
        return false;
    return isNumber(val)
        && !isNaN(val)
        && val % 1 !== 0;
}
exports.isDouble = isDouble;
/**
 * returns true if val is NaN
 * @see {Number.isNaN}
 */
function isNaN(val) {
    return Number.isNaN(val);
}
exports.isNaN = isNaN;
/**
 * returns true if val is finite
 * @see {Number.isFinite}
 */
function isFinite(val) {
    return Number.isFinite(val);
}
exports.isFinite = isFinite;
