// is-type-of/dist/commonjs/types/std.js
"use strict";
// Javascript Standard Object
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPromiseLike = exports.isPromise = exports.isGenerator = exports.isError = exports.isDate = exports.isRegExp = exports.isClass = exports.isObject = exports.isAsyncGeneratorFunction = exports.isAsyncFunction = exports.isGeneratorFunction = exports.isFunction = exports.isArray = void 0;
const node_util_1 = require("node:util");
const util_js_1 = require("./util.js");
const debug = (0, node_util_1.debuglog)('is-type-of:std');
// Array
/**
 * returns true if val is array, it won't check items of array
 */
function isArray(val) {
    return Array.isArray(val);
}
exports.isArray = isArray;
// Function
/**
 * returns true if val is function
 */
function isFunction(val) {
    return typeof val === 'function';
}
exports.isFunction = isFunction;
/**
 * returns true if val is generator function
 */
function isGeneratorFunction(val) {
    return node_util_1.types.isGeneratorFunction(val);
}
exports.isGeneratorFunction = isGeneratorFunction;
/**
 * returns true if val is async function
 */
function isAsyncFunction(val) {
    return node_util_1.types.isAsyncFunction(val);
}
exports.isAsyncFunction = isAsyncFunction;
/**
 * returns true if val is async generator function
 */
function isAsyncGeneratorFunction(val) {
    if (!isFunction(val))
        return false;
    return val.constructor && val.constructor.name === 'AsyncGeneratorFunction';
}
exports.isAsyncGeneratorFunction = isAsyncGeneratorFunction;
// Object
/**
 * returns true if val is object
 */
function isObject(val) {
    return typeof val === 'object' && val !== null;
}
exports.isObject = isObject;
/**
 * returns true if val is class
 * *Note:* "class" is supported in ECMAScript 6, and if the code is using some compiler or transpiler, the checking might fail
 */
function isClass(val) {
    if (!isFunction(val))
        return false;
    const fnStr = Function.prototype.toString.call(val);
    if (/^class[\s{]/.test(fnStr))
        return true;
    debug('[isClass] fnStr: %s', fnStr);
    const fnBody = fnStr.replace(/^[^{]*{\s*/, '').replace(/\s*}[^}]*$/, '');
    debug('[isClass] fnBody: %s', fnBody);
    return (/classCallCheck\(/.test(fnBody) || /TypeError\("Cannot call a class as a function"\)/.test(fnBody));
}
exports.isClass = isClass;
/**
 * returns true if val is regular expression
 */
function isRegExp(val) {
    return (0, util_js_1.isInstanceOf)(val, RegExp);
}
exports.isRegExp = isRegExp;
/**
 * returns true if val is instance of Date
 */
function isDate(val) {
    return (0, util_js_1.isInstanceOf)(val, Date);
}
exports.isDate = isDate;
/**
 * returns true if val is instance of Error
 */
function isError(val) {
    return (0, util_js_1.isInstanceOf)(val, Error);
}
exports.isError = isError;
/**
 * returns true if val is generator
 */
function isGenerator(val) {
    return node_util_1.types.isGeneratorObject(val);
}
exports.isGenerator = isGenerator;
/**
 * returns true if val is promise
 */
function isPromise(val) {
    return node_util_1.types.isPromise(val);
}
exports.isPromise = isPromise;
/**
 * returns true if val is like promise, if the object has `then` property, the checking will pass.
 */
function isPromiseLike(val) {
    if (isPromise(val))
        return true;
    if (!isObject(val))
        return false;
    return (0, util_js_1.hasOwnProperty)(val, 'then') && typeof val.then === 'function';
}
exports.isPromiseLike = isPromiseLike;
