// @ctrip/shark-sdk/lib/metric/index.js
'use strict'
const soa = require('../soa')
const addon = require('../store/addon')
const foundation = require('../common/foundation')
const pkg = require('../../package.json')

class Metric {
  constructor () {
    this.head = {
      appId: foundation.appID,
      version: pkg.version,
      sdk: 'Node',
      ip: foundation.clientIp
    }
  }

  reportMetric () {
    const dimensions = addon.dumpMetric()
    if (dimensions.length) {
      return soa.logging.reportMetric({ dimensions, head: this.head })
    }
    return Promise.resolve()
  }

  reportUsedKeys () {
    const recentUsedKey = addon.dumpReportUsedKey()
    if (recentUsedKey.length) {
      return soa.logging.reportRecentUsedKey({ recentUsedKey, head: this.head })
    }
    return Promise.resolve()
  }

  reportUndefinedKeys () {
    const keyLog = addon.dumpUndefinedKey()
    if (keyLog.length) {
      return soa.logging.reportKeyNotFound({ keyLog, head: this.head })
    }
    return Promise.resolve()
  }
}

module.exports = new Metric()
