// @ctrip/shark-sdk/lib/common/constants.js
'use strict'
const pkg = require('../../package.json')

module.exports = {

  /**
   * sdk version
   */
  SDK_VERSION: pkg.version,

  /**
   * npm package version
   */
  NPM_VERSION: pkg.version,

  /**
   * 托底语种
   */
  BASE_LOCALE: 'en-US',

  /**
   * 原文语种
   */
  ORIGIN_LOCALE: 'ORIGIN',

  /**
   * soa默认请求超时时间
   */
  SOA_TIMEOUT_MILLISECONDS: 10000,

  /**
   * file默认请求超时时间
   */
  FILE_TIMEOUT_MILLISECONDS: 30000,

  /**
   * baseLocaleConfig filename
   */
  FILE_CONFIG_FILENAME: 'config.json',

  /**
   * meta filename, include timestamp, ignite
   */
  FILE_META_FILENAME: 'meta.json',

  /**
   * 点火超时时间
   */
  IGNITION_TIMEOUT_MILLISECONDS: {
    MASTER: 50000,
    WAIT_SLAVE_SETUP: 3000,
    SLAVE_CONSUME: 30000,
    SLAVE: 50000
  },

  /**
   * file name regexp
   */
  TRANS_FILENAME_MATCH: /^(?<innerAppID>\d+)_(?<timestamp>\d+)\.json$/,

  /**
   * L10N appid
   */
  L10N_APPID: 6002,

  /**
   * Timer intervals
   */
  INTERVALS: {
    DEFAULT: 600000,
    METRIC: 600000,
    TALOS_INCREMENTAL: 60000,
    TALOS_FULL: 1800000,
    SOURCE_URL: {
      SHA: 'https://download2.ctripcorp.com/sharkappidtransv2/%s.zip',
      'SIN-AWS': 'https://ibu-sin-shark.s3.ap-southeast-1.amazonaws.com/apptrans/pro/v2/%s.zip',
      'FRA-AWS': 'https://ibu-fra-shark.s3.eu-central-1.amazonaws.com/apptrans/pro/v2/%s.zip'
    },
    REPORT_KEYS: () => Math.ceil(3600000 * Math.random()) + 3600000
  },

  /**
   * talos update mode
   */
  UPDATE_MODE: {
    INCREMENTAL: 0,
    FULL: 1
  },

  /**
   * group enum
   */
  GROUPS: {
    CTRIP: 'ctrip',
    TRIP: 'trip'
  },

  /**
   * ipc message types
   */
  IPC_MESSAGE: {
    PROTO_CONTENTS: '0',
    PROTO_PART_CONTENTS: '1',
    JSON_CONTENTS: '2',
    JSON_PART_CONTENTS: '3',
    TALOS_TRANS: '4',
    TALOS_FULL_TRANS: '5',
    TALOS_BASE_LOCALE: '6'
  },

  /**
   * enable shark plugin cookie name
   */
  PLUGIN_COOKIE: 'shark_plugin',

  /**
   * qconfig
   */
  QCONFIG: {
    FILENAME: 'shark-common.properties',
    USE_ORIGIN: 'debug.use.origin'
  },

  /**
   * context default
   */
  CONTEXT_DEFAULT: {
    GROUP: 'trip',
    LOCALE: 'en-US'
  },

  /**
   * Process Role
   */
  ROLE: {
    MASTER: 0,
    SLAVE: 1
  },

  /**
   * Cat Category Name
   */
  CAT_NAME: 'Shark.SDK',

  /**
   * Cat Span Keys
   */
  CATSPAN_TYPE: {
    IGNITION: 'Ignition',
    UPDATE: 'Update',
    FULL_UPDATE: 'FullUpdate'
  },

  /**
   * Cat Event Keys
   */
  CATEVENT_TYPE: {
    RPC: 'Shark.Rpc',
    RPC_END: 'Shark.Rpc.End',
    IPC: 'Shark.Ipc',
    FULL_DATA: 'Shark.Full.Data',
    IGNITE: 'Shark.Ignite',
    REQUIRE_ERROR: 'Shark.Require.Error'
  },

  /**
   * I18n Update Events
   */
  UPDATE_EVENTS: {
    KEYS: 'keys',
    BASE_LOCALE: 'base-locale'
  }
}
