// @ctrip/cargo-nodejs/test/context.test.js
'use strict'
const http = require('http')
const assert = require('assert')
const express = require('express')
const supertest = require('supertest')
const PORT = 3000

/**
 * @param {import('../index')} cargo
 */
module.exports = (cargo) => {

  let server, client

  it('start webserver', done => {
    const app = express()
    app.use(cargo.middlewares.express())
    app.get('/*', (req, res) => {
      res.json({
        group: req.cargoContext.getGroup(),
        site: req.cargoContext.getSite(),
        locale: req.cargoContext.getLocale(),
        strictLocale: req.cargoContext.getStrictLocale(),
        language: req.cargoContext.getLanguage(),
        currency: req.cargoContext.getCurrency(),
        htmlAttribue: req.cargoContext.toHTMLAttribute(),
        htmlAttribueString: req.cargoContext.toHTMLAttributeString(),
        getLocaleCookie: req.cargoContext.getCookie('locale'),
      })
    })
    server = http.Server(app)
    server.listen(PORT)
    client = supertest(server)
    done()
  })

  it(`curl localhost
        -H 'Host: www.dev.qa.nt.tripcorp.com'
  `, done => {
    client.get('/').set('Host', 'www.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'EN',
          locale: 'en-US',
          strictLocale: 'en-US',
          language: 'en',
          currency: 'USD',
          htmlAttribue: {
            'data-cargo': 'locale:en-US,language:en,currency:USD,contextType:online,site:EN,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-US,language:en,currency:USD,contextType:online,site:EN,group:Trip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl localhost
        -H 'Host: hk.dev.qa.nt.tripcorp.com'
  `, done => {
    client.get('/').set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'zh-HK',
          strictLocale: 'zh-HK',
          language: 'hk',
          currency: 'HKD',
          htmlAttribue: {
            'data-cargo': 'locale:zh-HK,language:hk,currency:HKD,contextType:online,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:zh-HK,language:hk,currency:HKD,contextType:online,site:HK,group:Trip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl localhost
        -H 'Host: english.dev.qa.nt.ctripcorp.com'
  `, done => {
    client.get('/').set('Host', 'english.dev.qa.nt.ctripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'EN',
          locale: 'en-US',
          strictLocale: 'en-US',
          language: 'en',
          currency: 'USD',
          htmlAttribue: {
            'data-cargo': 'locale:en-US,language:en,currency:USD,contextType:online,site:EN,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-US,language:en,currency:USD,contextType:online,site:EN,group:Ctrip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost?locale=en-hk&curr=CNY'
        -H 'Host: hk.dev.qa.nt.tripcorp.com'
  `, done => {
    client.get('/?curr=CNY&locale=en-hk')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        console.log(setCookie)
        assert.ok(
          setCookie.indexOf('ibulanguage=EN') !== -1 &&
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost?locale=en_hk&curr=CNY'
        -H 'Host: hk.dev.qa.nt.tripcorp.com'
  `, done => {
    client.get('/?curr=CNY&locale=en_hk')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibulanguage=EN') !== -1 &&
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost?locale=en-hk&curr=CNY'
        -H 'Host: www.dev.qa.nt.ctripcorp.com.hk'
  `, done => {
    client.get('/?curr=CNY&locale=en-hk')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Ctrip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost?locale=en-hk&curr=CNY&curr=USD'
        -H 'Host: www.dev.qa.nt.ctripcorp.com.hk'
  `, done => {
    client.get('/?curr=CNY&curr=USD&locale=en-hk')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Ctrip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost'
        -H 'Host: hk.dev.qa.nt.tripcorp.com'
        -H 'Cookie: ibulocale=en-hk; cookiePricesDisplayed=CNY'
  `, done => {
    client.get('/')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .set('Cookie', 'ibulocale=en-hk; cookiePricesDisplayed=CNY')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip"',
          getLocaleCookie: 'en-hk'
        })
        done()
      })
  })

  it(`curl 'localhost'
        -H 'Host: www.dev.qa.nt.ctripcorp.com.hk'
        -H 'Cookie: ibulocale=en-hk; cookiePricesDisplayed=USD'
  `, done => {
    client.get('/')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .set('Cookie', 'ibulocale=en-hk; cookiePricesDisplayed=USD')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'USD',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:USD,contextType:online,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:USD,contextType:online,site:HK,group:Ctrip"',
          getLocaleCookie: 'en-hk'
        })
        done()
      })
  })

  it(`curl 'localhost'
        -H 'Host: hk.dev.qa.nt.tripcorp.com'
        -H 'Locale: en-hk'
        -H 'Currency: CNY'
  `, done => {
    client.get('/')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .set('Locale', 'en-hk')
      .set('Currency', 'cny')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'CNY',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:CNY,contextType:online,site:HK,group:Trip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it(`curl 'localhost'
        -H 'Host: www.dev.qa.nt.ctripcorp.com.hk'
        -H 'Locale: en-hk'
        -H 'Currency: USD'
  `, done => {
    client.get('/')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .set('Locale', 'en-hk')
      .set('Currency', 'usD')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibulocale=en_hk') !== -1 &&
          setCookie.indexOf('cookiePricesDisplayed=USD') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'en-HK',
          strictLocale: 'en-HK',
          language: 'en',
          currency: 'USD',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:en,currency:USD,contextType:online,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:en,currency:USD,contextType:online,site:HK,group:Ctrip"',
          getLocaleCookie: ''
        })
        done()
      })
  })

  it('stop webserver', function (done) {
    server.close(done)
  })
}
