// @ctrip/cargo-nodejs/test/context-h5.test.js
'use strict'
const http = require('http')
const assert = require('assert')
const express = require('express')
const supertest = require('supertest')
const PORT = 3001

/**
 * @param {import('../index')} cargo
 */
module.exports = (cargo) => {

  let server, client

  it('start webserver', done => {
    const app = express()
    app.use(cargo.middlewares.express(cargo.CONTEXT_TYPE.H5))
    app.get('/*', (req, res) => {
      res.json({
        group: req.cargoContext.getGroup(),
        site: req.cargoContext.getSite(),
        locale: req.cargoContext.getLocale(),
        currency: req.cargoContext.getCurrency(),
        domain: req.cargoContext.getDomain(),
        htmlAttribue: req.cargoContext.toHTMLAttribute(),
        htmlAttribueString: req.cargoContext.toHTMLAttributeString()
      })
    })
    server = http.Server(app)
    server.listen(PORT)
    client = supertest(server)
    done()
  })

  it(`curl www.dev.qa.nt.tripcorp.com`, done => {
    client.get('/').set('Host', 'www.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'EN',
          locale: 'en-US',
          currency: 'USD',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:en-US,language:en,currency:USD,contextType:h5,site:EN,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-US,language:en,currency:USD,contextType:h5,site:EN,group:Trip"'
        })
        done()
      })
  })

  it(`curl hk.dev.qa.nt.tripcorp.com`, done => {
    client.get('/').set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'zh-HK',
          currency: 'HKD',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:zh-HK,language:hk,currency:HKD,contextType:h5,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:zh-HK,language:hk,currency:HKD,contextType:h5,site:HK,group:Trip"'
        })
        done()
      })
  })

  it(`curl english.dev.qa.nt.ctripcorp.com`, done => {
    client.get('/').set('Host', 'english.dev.qa.nt.ctripcorp.com')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'EN',
          locale: 'en-US',
          currency: 'USD',
          domain: 'qa.nt.ctripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:en-US,language:en,currency:USD,contextType:h5,site:EN,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:en-US,language:en,currency:USD,contextType:h5,site:EN,group:Ctrip"'
        })
        done()
      })
  })

  it(`curl hk.dev.qa.nt.tripcorp.com?locale=en-hk&curr=CNY`, done => {
    client.get('/?curr=CNY&locale=en-hk')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=en-hk') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          currency: 'CNY',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip"'
        })
        done()
      })
  })

  it(`curl jp.dev.qa.nt.tripcorp.com?locale=hk&curr=CNY`, done => {
    client.get('/?curr=CNY&locale=hk')
      .set('Host', 'jp.dev.qa.nt.tripcorp.com')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=ja-jp') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'JP',
          locale: 'ja-JP',
          currency: 'CNY',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:ja-JP,language:jp,currency:CNY,contextType:h5,site:JP,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:ja-JP,language:jp,currency:CNY,contextType:h5,site:JP,group:Trip"'
        })
        done()
      })
  })

  it(`curl english.dev.qa.nt.ctripcorp.com?lang=hk&currency=CNY`, done => {
    client.get('/?curr=CNY&lang=hk')
      .set('Host', 'english.dev.qa.nt.ctripcorp.com')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=zh-hk') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'EN',
          locale: 'zh-HK',
          currency: 'CNY',
          domain: 'qa.nt.ctripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:zh-HK,language:hk,currency:CNY,contextType:h5,site:EN,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:zh-HK,language:hk,currency:CNY,contextType:h5,site:EN,group:Ctrip"'
        })
        done()
      })
  })

  it(`curl 'www.dev.qa.nt.ctripcorp.com.hk?lang=KR&curr=CNY&curr=USD'`, done => {
    client.get('/?curr=CNY&curr=USD&lang=KR')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=ko-kr') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'ko-KR',
          currency: 'CNY',
          domain: 'qa.nt.ctripcorp.com.hk',
          htmlAttribue: {
            'data-cargo': 'locale:ko-KR,language:kr,currency:CNY,contextType:h5,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:ko-KR,language:kr,currency:CNY,contextType:h5,site:HK,group:Ctrip"'
        })
        done()
      })
  })

  it(`curl 'hk.dev.qa.nt.tripcorp.com' -H 'Cookie: ibu_h5_local=en-hk; ibu_h5_curr=CNY'`, done => {
    client.get('/')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .set('Cookie', 'ibu_h5_local=en-hk; ibu_h5_curr=CNY')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          currency: 'CNY',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip"'
        })
        done()
      })
  })

  it(`curl 'localhost'
        -H 'Host: www.dev.qa.nt.ctripcorp.com.hk'
        -H 'Cookie: ibu_h5_local=en-hk; ibu_h5_curr=USD'
  `, done => {
    client.get('/')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .set('Cookie', 'ibu_h5_local=zh-hk; ibu_h5_curr=USD')
      .expect(200)
      .end((err, res) => {
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'zh-HK',
          currency: 'USD',
          domain: 'qa.nt.ctripcorp.com.hk',
          htmlAttribue: {
            'data-cargo': 'locale:zh-HK,language:hk,currency:USD,contextType:h5,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:zh-HK,language:hk,currency:USD,contextType:h5,site:HK,group:Ctrip"'
        })
        done()
      })
  })

  it(`curl 'hk.dev.qa.nt.tripcorp.com' -H 'Locale: en-hk' -H 'Currency: CNY'`, done => {
    client.get('/')
      .set('Host', 'hk.dev.qa.nt.tripcorp.com')
      .set('Locale', 'en-hk')
      .set('Currency', 'cny')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=en-hk') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=CNY') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Trip',
          site: 'HK',
          locale: 'en-HK',
          currency: 'CNY',
          domain: 'qa.nt.tripcorp.com',
          htmlAttribue: {
            'data-cargo': 'locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip'
          },
          htmlAttribueString: 'data-cargo="locale:en-HK,language:enHK,currency:CNY,contextType:h5,site:HK,group:Trip"'
        })
        done()
      })
  })

  it(`curl 'www.dev.qa.nt.ctripcorp.com.hk' -H 'Locale: en-hk' -H 'Currency: USD'`, done => {
    client.get('/')
      .set('Host', 'www.dev.qa.nt.ctripcorp.com.hk')
      .set('Locale', 'en-hk')
      .set('Currency', 'usD')
      .expect(200)
      .end((err, res) => {
        const setCookie = res.get('set-cookie').join('\n')
        assert.ok(
          setCookie.indexOf('ibu_h5_local=zh-hk') !== -1 &&
          setCookie.indexOf('ibu_h5_curr=USD') !== -1
        )
        assert.deepStrictEqual(res.body, {
          group: 'Ctrip',
          site: 'HK',
          locale: 'zh-HK',
          currency: 'USD',
          domain: 'qa.nt.ctripcorp.com.hk',
          htmlAttribue: {
            'data-cargo': 'locale:zh-HK,language:hk,currency:USD,contextType:h5,site:HK,group:Ctrip'
          },
          htmlAttribueString: 'data-cargo="locale:zh-HK,language:hk,currency:USD,contextType:h5,site:HK,group:Ctrip"'
        })
        done()
      })
  })

  it('stop webserver', function (done) {
    server.close(done)
  })
}
