// @ctrip/shark-l10n/l10n.js
'use strict'
import dayjs from 'dayjs'
import { L10n, ROUNDING_MODE, Unit, UnitCollection } from './lib'

var instance = new L10n(window.i18n_6002 || {})

var l10n = {

  l10n: window.i18n_6002 || {},

  dayjs,

  DateTimeFormat: function (timezone) {
    return instance.setI18n(l10n.l10n).DateTimeFormat(timezone)
  },

  DurationFormat: function (locale) {
    return instance.setI18n(l10n.l10n).DurationFormat(locale)
  },

  MeasurementFormat: function () {
    return instance.setI18n(l10n.l10n).MeasurementFormat()
  },

  NameFormat: function () {
    return instance.setI18n(l10n.l10n).NameFormat();
  },

  NumberFormat: function () {
    return instance.setI18n(l10n.l10n).NumberFormat()
  },

  PhoneFormat: function () {
    return instance.setI18n(l10n.l10n).PhoneFormat()
  },

  getPlural: function (i18n, key, locale, num) {
    return instance.setI18n(i18n).getPlural(key, locale, num);
  },

  ROUNDING_MODE: ROUNDING_MODE,

  Unit: Unit,

  UnitCollection: UnitCollection
}

window.l10n = l10n

export default l10n
