// @ctrip/shark-l10n/es/index.js
import { DateTimeFormatter } from './datetime';
import { DurationFormatter } from './duration';
import { MeasurementFormatter } from './measurement';
import { NameFormatter } from './name';
import { NumberFormatter } from './number';
import { PhoneFormatter } from './phone';
import { getPlural } from './plural';
export * from './types';
var L10n = (function () {
    function L10n(i18n) {
        this.setI18n(i18n);
    }
    L10n.prototype.setI18n = function (i18n) {
        this.i18n = i18n;
        return this;
    };
    L10n.prototype.DateTimeFormat = function (timezone) {
        return new DateTimeFormatter(this.i18n, timezone);
    };
    L10n.prototype.DurationFormat = function (locale) {
        return new DurationFormatter(this.i18n, locale);
    };
    L10n.prototype.MeasurementFormat = function () {
        return new MeasurementFormatter(this.i18n);
    };
    L10n.prototype.NameFormat = function () {
        return new NameFormatter(this.i18n);
    };
    L10n.prototype.NumberFormat = function () {
        return new NumberFormatter(this.i18n);
    };
    L10n.prototype.PhoneFormat = function () {
        return new PhoneFormatter(this.i18n);
    };
    L10n.prototype.getPlural = function (key, locale, num) {
        return getPlural(this.i18n, key, locale, num);
    };
    return L10n;
}());
export { L10n };
export { DateTimeFormatter, DurationFormatter, MeasurementFormatter, NameFormatter, NumberFormatter, PhoneFormatter, getPlural, };
