// @ctrip/shark-l10n/es/types.js
export var CredentialType;
(function (CredentialType) {
    CredentialType["IDCardCredentialType"] = "IDCard";
    CredentialType["PassportCredentialType"] = "Passport";
    CredentialType["DrivingCardCredentialType"] = "DrivingCard";
    CredentialType["HomeReturnPermitCredentialType"] = "HRP";
    CredentialType["MTPCredentialType"] = "MTP";
    CredentialType["UsualCredentialType"] = "Usual";
    CredentialType["UsualNoMidNameCredentialType"] = "UsualNoMidName";
})(CredentialType || (CredentialType = {}));
export var ROUNDING_MODE;
(function (ROUNDING_MODE) {
    ROUNDING_MODE[ROUNDING_MODE["DOWN"] = 0] = "DOWN";
    ROUNDING_MODE[ROUNDING_MODE["HALF_UP"] = 1] = "HALF_UP";
    ROUNDING_MODE[ROUNDING_MODE["HALF_EVEN"] = 2] = "HALF_EVEN";
    ROUNDING_MODE[ROUNDING_MODE["UP"] = 3] = "UP";
})(ROUNDING_MODE || (ROUNDING_MODE = {}));
export var Unit;
(function (Unit) {
    Unit["Kilometer"] = "Kilometer";
    Unit["Mile"] = "Mile";
    Unit["SquareMeter"] = "SquareMeter";
    Unit["SquareFeet"] = "SquareFeet";
    Unit["Celsius"] = "Celsius";
    Unit["Fahrenheit"] = "Fahrenheit";
    Unit["Meter"] = "Meter";
    Unit["Feet"] = "Feet";
    Unit["Kilogram"] = "Kilogram";
    Unit["Pound"] = "Pound";
    Unit["Centimeter"] = "Centimeter";
    Unit["Inch"] = "Inch";
})(Unit || (Unit = {}));
export var UnitCollection;
(function (UnitCollection) {
    UnitCollection["Unknown"] = "Unknown";
    UnitCollection["MetricUnit"] = "MetricUnit";
    UnitCollection["ImperialUnit"] = "ImperialUnit";
})(UnitCollection || (UnitCollection = {}));
export var KEY_PREFERENCES_MEASUREMENT = 'key.l10n.unit.measurement';
export var KEY_PREFERENCES_TEMPERATURE = 'key.l10n.unit.temperature';
