// @ctrip/node-vampire-standard-soa-client/libs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoaClientAgent = exports.SoaClientDirectApi = exports.SoaClientWebApi = exports.CLIENT_TYPE = void 0;
// 用K_REVISION判断是否是发布环境，本地和MARS无需开启
const IS_FAAS = process.env.K_REVISION && process.env.FUNCTION_ID && process.env.FUNCTION_ID[0] === '3';
const USE_MESH_SIDECAR = process.env.MESH_SIDECAR_INJECT === 'true' || IS_FAAS;
const soaMonitor_1 = require("./utils/soaMonitor");
Object.defineProperty(exports, "CLIENT_TYPE", { enumerable: true, get: function () { return soaMonitor_1.CLIENT_TYPE; } });
const _SoaClientWebApi = __importStar(require("./soaWebApi"));
const _SoaClientDirectApi = __importStar(require("./soaDirectApi"));
const soaMeshApi_1 = require("./soaMeshApi");
exports.SoaClientWebApi = USE_MESH_SIDECAR ? soaMeshApi_1.SoaClientMeshApi : _SoaClientWebApi.SoaClientWebApi;
exports.SoaClientDirectApi = USE_MESH_SIDECAR ? soaMeshApi_1.SoaClientMeshApi : _SoaClientDirectApi.SoaClientDirectApi;
const SOAAgentCache = {};
function SoaClientAgent(serviceId, opt = {}) {
    let cacheKey = '';
    if (typeof serviceId === 'object') {
        opt = serviceId;
        cacheKey = opt.serviceId || opt.serviceCode;
    }
    else {
        opt.serviceId = serviceId;
        cacheKey = serviceId;
    }
    if (opt.soaServerUrl) {
        cacheKey += '-' + opt.soaServerUrl;
    }
    if (opt.subEnv) {
        cacheKey += '-' + opt.subEnv;
    }
    let soaClientApi = null;
    if (SOAAgentCache[cacheKey]) {
        soaClientApi = SOAAgentCache[cacheKey];
    }
    else {
        const SoaClientApi = opt.soaClientType === soaMonitor_1.CLIENT_TYPE.WebApi ? exports.SoaClientWebApi : exports.SoaClientDirectApi;
        soaClientApi = SOAAgentCache[cacheKey] = new SoaClientApi(opt);
    }
    return new Proxy(soaClientApi, {
        get: (target, key) => {
            return function (args) {
                return target.init().catch((err) => {
                    delete SOAAgentCache[cacheKey];
                    return Promise.reject(err);
                }).then(() => {
                    const _args = Object.assign({
                        method: key
                    }, args || {
                        args: {}
                    });
                    return soaClientApi.invoke(_args);
                });
            };
        }
    });
}
exports.SoaClientAgent = SoaClientAgent;
